### ISSUE TITLE
[Question] Best way to shard model with more than one EmbeddingCollection/EmbeddingBagCollection

### ISSUE THREAD

**Issue by shijieliu** [meta-pytorch/torchrec#1388]
Hi experts, I have some questions about how to correctly configure TorchRec to achieve best performance in my use case.

1. What's the differences between `EmbeddingBagCollection`/`EmbeddingCollection`? If the pooling_factor is 1, which one is prefered to use? 
2. When I have tables with different embedding_dim, is it required to group them in different `EmbeddingBagCollection`/`EmbeddingCollection`?
For example, if I have 2 tables whose embedding_dim is 8 and 16, I need to create 2 `EmbeddingBagCollection` or 2 `EmbeddingCollection`.
3. I am using `nn.ModuleList` to wrap `EmbeddingCollection`/`EmbeddingBagCollection` inside a nn.Module, the example is showed as followed. Is it the correct way? Will this affect the `EmbeddingShardingPlanner`? To be more specific, in this way, will the `EmbeddingShardingPlanner` consider all `EmbeddingCollection`/`EmbeddingBagCollection` possible sharding combinations?
```python

embedding_a = [
  EmbeddingBagConfig(
    feature_names=["a"],
    embedding_dim=8,
    ....
  )
]
ebc_a = EmbeddingBagCollection(tables=embedding_a, device='meta')

embedding_b = [
  EmbeddingBagConfig(
    feature_names=["b"],
    embedding_dim=16,
    ....
  )
]
ebc_b = EmbeddingBagCollection(tables=embedding_b, device='meta')

class SparseArch(nn.Module):
    def __init__(self, ebc):
        self.ebc = nn.ModuleList(ebc)

sparse_arch = SparseArch([ebc_a, ebc_b])
```
4. For `EmbeddingShardingPlanner`, how to configure it to achieve best performance? Right now I add some constriants about network_bandwidth and pooling_factor, is there other things I missed?
```python
num_table = 4
pooling_factor = [1, 1, 10, 10]
dict_const = {}
for i in range(num_table):
    const = ParameterConstraints()
    const.pooling_factors = pooling_factor[i]
    const.enforce_hbm = True
    dict_const['t_table_' + str(i)] = const

planner = EmbeddingShardingPlanner(
  topology=Topology(
      local_world_size=get_local_size(),
      world_size=dist.get_world_size(),
      compute_device=device.type,
      hbm_cap=80 * 1024 * 1024 * 1024,
      intra_host_bw=300e9,
      inter_host_bw=25e9,
  ),
  constraints=dict_const,
  batch_size=args.batch_size_per_gpu,
  # # If experience OOM, increase the percentage. see
  # # https://pytorch.org/torchrec/torchrec.distributed.planner.html#torchrec.distributed.planner.storage_reservations.HeuristicalStorageReservation
  storage_reservation=HeuristicalStorageReservation(percentage=0.05),
  debug=True
)
plan = planner.collective_plan(
    train_model, get_default_sharders(), dist.GroupMember.WORLD
)
```

**Comment by henrylhtsang** [comment]
Hi, let me try to answer the questions, and maybe you can ask follow-ups.

> 1. What's the differences between EmbeddingBagCollection/EmbeddingCollection? If the pooling_factor is 1, which one is prefered to use?

Its the same difference between nn.EmbeddingBag and nn.Embedding: In embeddingbag, there is pooling (sum or mean) happening while in Embedding there is no pooling. If your data is single hot probably you can just use Embedding. But both should work.

> 2. When I have tables with different embedding_dim, is it required to group them in different EmbeddingBagCollection/EmbeddingCollection?

No need to separate them.

> 3. I am using nn.ModuleList to wrap EmbeddingCollection/EmbeddingBagCollection inside a nn.Module, the example is showed as followed. Is it the correct way? Will this affect the EmbeddingShardingPlanner? To be more specific, in this way, will the EmbeddingShardingPlanner consider all EmbeddingCollection/EmbeddingBagCollection possible sharding combinations?

I think it should be fine. In both model_parallel and planner, we will look at the child modules and check what type they are, so it should be fine.
```
named_modules_queue = [("", module)]
        while named_modules_queue:
            child_path, child_module = named_modules_queue.pop()
            sharder_key = sharder_name(type(child_module))
            sharder = self._sharder_map.get(sharder_key, None)
            if not sharder:
                for n, m in child_module.named_children():
                    if child_path != "":
                        named_modules_queue.append((child_path + "." + n, m))
                    else:
                        named_modules_queue.append((n, m))
                continue
```

> 4. For EmbeddingShardingPlanner, how to configure it to achieve best performance? Right now I add some constriants about network_bandwidth and pooling_factor, is there other things I missed?

I would say, give as much information and constraints through **ParameterConstraints**, including sharding type, compute kernel, batch_size, min_partition (when using column_wise).

Edit: format

**Comment by shijieliu** [comment]
Hi @henrylhtsang Thanks for your quick response!

> What's the differences between EmbeddingBagCollection/EmbeddingCollection? If the pooling_factor is 1, which one is prefered to use?
Its the same difference between nn.EmbeddingBag and nn.Embedding: In embeddingbag, there is pooling (sum or mean) happening while in Embedding there is no pooling. If your data is single hot probably you can just use Embedding. But both should work.

Yes I know it works functionally for single hot with both `nn.EmbeddingBag` and `nn.Embedding`. And I would like to know if there any performance difference?

> When I have tables with different embedding_dim, is it required to group them in different EmbeddingBagCollection/EmbeddingCollection?
No need to separate them.

Thanks. I tried `EmbeddingBagCollection`. Looks it support different embedding_dim. But for `EmbeddingCollection` I get error `All tables in a EmbeddingCollection are required to have same embedding dimension.`. So I suppose `EmbeddingCollection` does not support different embedding_dim.

About your suggestion, I tried to add more constrains. And try to see with different sharding types what's the sharding result will be. My finding is as follows:
1. for `row_wise`, table will be sharded on all gpus
2. for `table_row_wise`, table will be sharded on `local_world_size` gpus.
3. for `table_wise`, table will be sharded on single gpu
4. for `table_column_wise`, I get duplicated ranks. Is this expected?
```python
sharding_type='table_column_wise', compute_kernel='fused', ranks=[112, 113, 114, 115, 116, 117, 118, 119, 112, 113, 114, 115, 116, 117, 118, 119, 112, 113, 114, 115, 116, 117, 118, 119, 112, 113, 114, 115, 116, 117, 118, 119]
```
6. for `column_wise`, the ranks are distributed in an interleaving way. Does this mean I have to use `column_wise` in different nodes?
```
sharding_type='column_wise', compute_kernel='fused', ranks=[124, 116, 108, 100, 92, 84, 76, 68, 60, 52, 44, 36, 28, 20, 12, 4, 125, 117, 109, 101, 93, 85, 77, 69, 61, 53, 45, 37, 29, 21, 13, 5]
```
Could you help answer and confirm? Thanks!

BTW, maybe there is a bug in  [this line](https://github.com/pytorch/torchrec/blob/200e2dfb3695624ab29ce2c7b39bd2a74f571b4f/torchrec/distributed/planner/enumerators.py#L185) and [this line](https://github.com/pytorch/torchrec/blob/200e2dfb3695624ab29ce2c7b39bd2a74f571b4f/torchrec/distributed/planner/enumerators.py#L208). It should be
```
constrained_sharding_types: List[str] = [v.value for v in constraints.sharding_types]
constrained_compute_kernels: List[str] = [v.value for v in constraints.compute_kernels]
```
Please check. Thanks!

**Comment by henrylhtsang** [comment]
> Yes I know it works functionally for single hot with both nn.EmbeddingBag and nn.Embedding. And I would like to know if there any performance difference?

I haven't done much work with EmbeddingCollection, so can't say for sure. But I imagine they are similar, since they both call the same fbgemm kernel.

> Thanks. I tried EmbeddingBagCollection. Looks it support different embedding_dim. But for EmbeddingCollection I get error All tables in a EmbeddingCollection are required to have same embedding dimension.. So I suppose EmbeddingCollection does not support different embedding_dim.

Okay didn't know that.

> About your suggestion, I tried to add more constrains. And try to see with different sharding types what's the sharding result will be. My finding is as follows:

Yeah your results look about right. Rowwise shard on all the gpus, so number of shards is the number of gpus. Table wise means not sharding them and only placing them into a single GPU. Column wise means sharding them into a certain number of shards, specified by your min_partition (e.g., if table dim is 800 and min_partition is 100, then you will get 8 shards). When you have more shards than gpus, some shards will end up in the same gpu. Now, table_column_wise and table_row_wise mean column_wise and row_wise, but restricted into the same host. 

> the ranks are distributed in an interleaving way. 

You can take a look at partitioners.py, where we also take `rank % local_world_size` into account. 

> BTW, maybe there is a bug in [this line](https://github.com/pytorch/torchrec/blob/200e2dfb3695624ab29ce2c7b39bd2a74f571b4f/torchrec/distributed/planner/enumerators.py#L185) and [this line](https://github.com/pytorch/torchrec/blob/200e2dfb3695624ab29ce2c7b39bd2a74f571b4f/torchrec/distributed/planner/enumerators.py#L208). It should be

I don't think they were bugs? In ParameterConstraints, they are passed in as Optional[List[str]].

**Comment by shijieliu** [comment]
Hi @henrylhtsang Thanks for your help! Really appreciate!

After several days of debugging, I finally find the bad perf in my use case is because of `pin_memory` in `data loader` but not from torchrec. And I learned a lot from your answers about the sharding. I think I have no more questions in this issue.

> I don't think they were bugs? In ParameterConstraints, they are passed in as Optional[List[str]].
Yes you are right, I pass the enum instead of the str. That causes the problem.

**Comment by colin2328** [comment]
@shijieliu great! closing this issue then
